app/wasm/test/custom_msg_test.go

```diff
@@ -0,0 +1,469 @@
+					First: wasmbindings.Swap{
+						PoolId:   state.StarPool,
+						DenomIn:  "uosmo",
+						DenomOut: "ustar",
+					},
+					// Note: you must use empty array, not nil, for valid Rust JSON
+					Route: []wasmbindings.Step{},
+					Amount: wasmbindings.SwapAmountWithLimit{
+						ExactIn: &wasmbindings.ExactIn{
+							Input:     sdk.NewInt(12000000),
+							MinOutput: sdk.NewInt(555000000),
+						},
+					},
+				}
+			},
+			initFunds: sdk.NewInt64Coin("uosmo", 13000000),
+			expectErr: true,
+		},
+		"exact in: not enough funds to swap": {
+			msg: func(state BaseState) *wasmbindings.SwapMsg {
+				return &wasmbindings.SwapMsg{
+					First: wasmbindings.Swap{
+						PoolId:   state.StarPool,
+						DenomIn:  "uosmo",
+						DenomOut: "ustar",
+					},
+					// Note: you must use empty array, not nil, for valid Rust JSON
+					Route: []wasmbindings.Step{},
+					Amount: wasmbindings.SwapAmountWithLimit{
+						ExactIn: &wasmbindings.ExactIn{
+							Input:     sdk.NewInt(12000000),
+							MinOutput: sdk.NewInt(5000000),
+						},
+					},
+				}
+			},
+			initFunds: sdk.NewInt64Coin("uosmo", 7000000),
+			expectErr: true,
+		},
+		"exact in: invalidPool": {
+			msg: func(state BaseState) *wasmbindings.SwapMsg {
+				return &wasmbindings.SwapMsg{
+					First: wasmbindings.Swap{
+						PoolId:   state.StarPool,
+						DenomIn:  "uosmo",
+						DenomOut: "uatom",
+					},
+					// Note: you must use empty array, not nil, for valid Rust JSON
+					Route: []wasmbindings.Step{},
+					Amount: wasmbindings.SwapAmountWithLimit{
+						ExactIn: &wasmbindings.ExactIn{
+							Input:     sdk.NewInt(12000000),
+							MinOutput: sdk.NewInt(100000),
+						},
+					},
+				}
+			},
+			initFunds: sdk.NewInt64Coin("uosmo", 13000000),
+			expectErr: true,
+		},
+
+		// FIXME: this panics in GAMM module !?! hits a known TODO
+		// https://github.com/osmosis-labs/osmosis/blob/a380ab2fcd39fb94c2b10411e07daf664911257a/osmomath/math.go#L47-L51
+		//"exact out: panics if too much swapped": {
+		//	msg: func(state BaseState) *wasmbindings.SwapMsg {
+		//		return &wasmbindings.SwapMsg{
+		//			First: wasmbindings.Swap{
+		//				PoolId:   state.StarPool,
+		//				DenomIn:  "uosmo",
+		//				DenomOut: "ustar",
+		//			},
+		//			// Note: you must use empty array, not nil, for valid Rust JSON
+		//			Route: []wasmbindings.Step{},
+		//			Amount: wasmbindings.SwapAmountWithLimit{
+		//				ExactOut: &wasmbindings.ExactOut{
+		//					MaxInput: sdk.NewInt(22000000),
+		//					Output:   sdk.NewInt(120000000),
+		//				},
+		//			},
+		//		}
+		//	},
+		//	initFunds: sdk.NewInt64Coin("uosmo", 15000000),
+		//	finalFunds: []sdk.Coin{
+		//		sdk.NewInt64Coin("uosmo", 3000000),
+		//		sdk.NewInt64Coin("ustar", 120000000),
+		//	},
+		//},
+
+		"exact out: simple swap works": {
+			msg: func(state BaseState) *wasmbindings.SwapMsg {
+				return &wasmbindings.SwapMsg{
+					First: wasmbindings.Swap{
+						PoolId:   state.AtomPool,
+						DenomIn:  "uosmo",
```

Comment:
Like here, OSMO -> ATOM works with ExactOut.

The above is a bug in the GAMM code (with existing TODO). I think it would be nice to fix, but that is out of scope of this PR